server/src/main.rs

```diff
@@ -85,18 +90,36 @@ impl LanguageServer for Backend {
     }
 
     async fn initialized(&self, _: InitializedParams) {
-        let workspace_uri = self.workspace_uri.lock().await;
+        // Update the sources and references directories
+        {
+            let workspace_uri = self.workspace_uri.lock().await;
+
+            // Fetch and set the sources directory
+            let sources_directory = self.get_sources_directory(&workspace_uri, None).await.ok();
+            let references_directories = self
+                .get_reference_directories(&workspace_uri, None)
+                .await
+                .ok();
+            let mut sources_uri = self.sources_uri.lock().await;
+            let mut references_uris = self.reference_uris.lock().await;
+            *sources_uri = sources_directory;
+            *references_uris = references_directories;
+        }
+
+        let mut shared_state_lock = self.shared_state.lock().await;
+
+        // Compile the Slice files and publish diagnostics
+        let (updated_state, options) = self.compile_slice_files().await;
+
+        shared_state_lock.compilation_state = updated_state;
+        shared_state_lock.compilation_options = options;
+
+        self.publish_diagnostics_for_all_files(&mut shared_state_lock)
+            .await;
```

Comment:
As I see it, the problem with `compile_slice_files` is that it has to deal with converting the extension state and settings into a `SliceOptions` struct for compiling. This adds unnecessary complexity to the function.

We can maintain a `sliceOptions` member, updating it when the config is updated or the extension is initialized, and then pass it as a parameter. It shouldn't be the job of `compile_slice_files` to compute the options. If we retain the current design of returning the compilation state, we can eliminate `compile_slice_files` and call `compile_from_options` directly. Otherwise, we should keep `compile_slice_files` and allow it to update the compilation state.

Reply:
I agree, 95% of the `compile_slice_files` is about computing options... How about separating it into 2 functions?
For cleanliness' sake?
```
async fn refresh_state() {
    let options = compute_options();
    let updated_state = slicec::compile_from_options(&option, ...);

    let mut shared_state_lock = self.shared_state.lock().await;
    // Update the options and shared_state here
}

async fn compute_options() -> SliceOptions {
    // This is mostly just `compile_slice_files`
    // It locks workspace_uri, sources_uri, and reference_uri
    // These locks are released at the the end of the function though.
}
```

> Otherwise, we should keep compile_slice_files and allow it to update the compilation state.

I still think it'd be nicer to let it set the compilation_state, just because the logic is repeated at each call site right now.